rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,524 @@
+actions one at a time, grouping similar endpoints into separate Github PRs.
+
+Within each PR, we can ensure that every usage of that endpoint is updated in a
+backwards compatible way across all Teleport clients (`tsh`, `tctl`, Teleport
+Connect, Teleport Web UI, Plugins and plugin guides).
+
+#### MVP Admin Actions Subset
+
+Again, this list is too long to implement all changes in one go. This change
+will likely take 1-2 major version cycles to complete. As such, the initial
+MVP goal will be limited to the following RBAC centric endpoints:
+
+- `GenerateUserCerts`, `GenerateHostCerts`
+- `CreateUser`, `UpdateUser`, `DeleteUser`
+- `SetAccessRequestState`, `SubmitAccessReview`
+- `UpsertTokenV2`, `CreateTokenV2`, `GenerateToken`, `DeleteToken`
+- `UpsertOIDCConnector`, `DeleteOIDCConnector`, `CreateOIDCAuthRequest`
+- `UpsertSAMLConnector`, `DeleteSAMLConnector`, `CreateSAMLAuthRequest`
+- `UpsertGithubConnector`, `DeleteGithubConnector`, `CreateGithubAuthRequest`
+- `SetClusterNetworkingConfig`, `ResetClusterNetworkingConfig`
+- `SetSessionRecordingConfig`, `ResetSessionRecordingConfig`
+- `SetAuthPreference`, `ResetAuthPreference`
+- `SetNetworkRestrictions`, `DeleteNetworkRestrictions`
+
+### MFA enforcement
+
+MFA verification will be required to carry out various administrative actions.
+This will be enforced by the Auth API Server for the specific endpoints which
+are considered [administrative actions](#administrative-actions).
+
+Each Admin Action API request will follow this flow:
+
+1. Create an Auth API client using the user's existing valid certificates.
+1. Use the client to retrieve an MFA challenge for the user with
+   the existing `rpc CreateAuthenticateChallenge`.
+1. Prompt the user to solve the MFA challenge with their MFA device.
+1. Send the resulting `MFAAuthenticateResponse` to the Auth Server as part of the
+   API request.
+1. Validate and consume the `MFAAuthenticateResponse` in the Auth Server to
+   authorize the request (in addition to normal identity-based authorization).
+
+Steps 1-3 are already possible with Teleport currently and used for various MFA
+features. Steps 4 and 5 will require some changes to the Auth API client and
+server respectively.
+
+#### Server changes
+
+For admin actions, the Auth server will validate MFA for each request using the
+`MFAAuthenticateResponse` passed in the request.
+
+If the request fails MFA verification, an access denied error will be returned
+to the client.
+
+```go
+// ErrAPIMFARequired is returned by AccessChecker when an API request
+// requires an MFA check.
+var ErrAdminActionMFARequired = trace.AccessDenied("administrative action requires MFA")
+```
+
+#### Client changes
+
+There are a few different ways that the Auth client can pass MFA verification with
+the Auth server, each with their own pros and cons:
+
+1. Add the `MFAAuthenticateResponse` argument to each administrative API endpoint.
```

Comment:
Sure thing, what do you think about `AugmentAuthentication`?

```proto
message AugmentAuthentication {
  MFAAuthenticateResponse mfa_authenticate_response = 1;
}
```

> I'd also suggest standardizing a field name for the challenge-response message, whatever it ends up being, so we can easily intercept for it among multiple RPC services.

I'm not sure if I follow. Do you mean that the wrapper above should be re-defined in each separate service with the standardized name and fields above?

Reply:
`AugmentAuthorization` (we already cleared authn, right)? `InBandAuthorization`? `AuthorizationMetadata`? `PerCallAuthorization`? (Something else?)

For the field name, I meant we should always use the same field name, so we can find it via proto-reflection (which is cheap and not like actual reflection) and intercept for it.

Eg:

```diff
message GenerateUserCerts {
  // ...

+ AuthorizationMetadata request_authorization_metadata = n; // Always use this field name!
}
```

(Names TBD, ofc.)

Reply:
Technically I think it's still authentication, right? We use MFA or Device Trust to increase our certainty of the user's identity, but it doesn't change what they are authorized to perform.

I think something like access requests would better fit the definition of authz augmentation. Let me know if I'm missing something.

Reply:
> The caveat is that we'll probably need either an allow or deny list for RPCs that require one check but not the other, but the upside is that enabling/disabling and even testing can be rather simple.

On second thought, maybe we should just keep these separate. Looping them together would look nice in the proto file but would actually add this complexity of not knowing if the specific service requests _actually_ requires MFA/Device trust.

The Device trust solution can still piggy back off of this in terms of implementation, as I imagine it will be largely identical.

Reply:
> Technically I think it's still authentication, right? (...)

On second (forth?) thought I agree with you. Authn it is.

> On second (third?) thought, maybe we should just keep these separate (...)

I was a bit on the fence because of that too. Let's wait to see if we are doing 1) or 2) - if we go with the latter I think we'll need to list applicable RPCs anyway, then bundling isn't much of a concern.